packages/core/data-layer.ts

```diff
@@ -3,20 +3,41 @@ import {Selector, Store, Unsubscribe} from "./model";
 export function subscribeWithSelector<S,K>(
     store: Store<S>,
     selector: Selector<S,K>,
-    cb: (state: K) => void
+    cb: (state: K) => void,
+    stateEquals: StateEquals<K>,
 ): Unsubscribe {
-    let prevValue = selector(
+    let prevState = selector(
         store.getState()
     );
     return store.subscribe(() => {
-        const newValue = selector(
+        const curState = selector(
             store.getState()
         );
-        if (newValue != prevValue) {
-            prevValue = newValue;
+        if (!stateEquals(prevState, curState)) {
+            prevState = curState;
             cb(
-                newValue,
+                curState,
             )
         }
     })
+}
+
+export interface StateEquals<K> {
+    (prevState: K, curState: K): boolean
+}
+
+export function shallowEquals<K>(obj1: K, obj2: K): boolean {
+    const obj1Keys = Object.keys(obj1);
+    const obj2Keys = Object.keys(obj2);
+    if (obj1Keys.length != obj2Keys.length) {
+        return false
+    }
+    for (const key in obj1) {
+        if (obj1.hasOwnProperty(key)) {
```

Comment:
`if (obj2.hasOwnProperty(key)) {` obj1 -> obj2 probably?

Reply:
what does it change? `obj2['notExistedKey'] == undefined` -> won't pass comparison, won't throw an error

Reply:
In outer loop you have iteration by all props  inside obj1, so, I don't understand why do you need to check it again in obj1, but we have a reason to check it in obj2, because we can't be sure that this pair key-values exists in obj2
`for (const key in obj1) {
  if (obj2.hasOwnProperty(key) && obj1[key] !== obj2[key]) {
    return false
  }
}
`

Reply:
it will return `true` for 
````
var a = {a: 1}, b = {b: 2}
````
because `b` does not have property `a` so `if` statement won't be executed